src/librustdoc/test.rs

```diff
@@ -422,11 +422,14 @@ impl Collector {
 
 impl DocFolder for Collector {
     fn fold_item(&mut self, item: clean::Item) -> Option<clean::Item> {
-        let pushed = match item.name {
-            Some(ref name) if name.is_empty() => false,
-            Some(ref name) => { self.names.push(name.to_string()); true }
-            None => false
-        };
+        let name = item.name.iter()
+                            .filter(|name| !name.is_empty())
+                            .cloned()
+                            .next()
+                            .or_else(|| typename_if_impl(&item));
+
+        let pushed = name.map(|name| self.names.push(name)).is_some();
```

Comment:
This is a little bit hard to follow, so perhaps this could use explicit control flow + `if let Some(...) = ...` to match what's going on here? Currently it's somewhat non-idiomatic to call `.iter()` on an option or rely on functions like `map` for side effects.